app/controllers/provider_interface/offer_changes_controller.rb

```diff
@@ -0,0 +1,159 @@
+      end
+    end
+
+  private
+
+    def render_providers
+      set_alternative_providers
+      render :edit_provider
+    end
+
+    def render_courses
+      set_alternative_courses
+      render :edit_course
+    end
+
+    def render_course_options
+      set_alternative_course_options
+      render :edit_course_option
+    end
+
+    def requires_provider_change_response_feature_flag
+      raise unless FeatureFlag.active?('provider_change_response')
+    end
+
+    def set_application_choice
+      @application_choice = GetApplicationChoicesForProviders.call(providers: current_provider_user.providers)
+        .find(params[:application_choice_id])
+    end
+
+    def set_change_offer_form
+      @change_offer_form = ProviderInterface::ChangeOfferForm.new application_choice: @application_choice,
+                                                                  provider_id: (provider.id if allowed_provider?),
+                                                                  course_id: course&.id,
+                                                                  course_option_id: course_option&.id
+    end
+
+    def allowed_provider?
+      current_provider_user.providers.include?(provider)
+    end
+
+    def provider
+      requested_provider || current_provider
+    end
+
+    def current_provider
+      @application_choice.offered_course.provider
+    end
+
+    def requested_provider
+      provider_id = change_offer_params[:provider_id]
+      Provider.find(provider_id) if provider_id.present?
+    end
+
+    def course
+      requested_course || current_course
+    end
+
+    def current_course
+      @application_choice.offered_course
+    end
+
+    def requested_course
+      course_id = change_offer_params[:course_id]
+      Course.find(course_id) if course_id.present?
+    end
+
+    def course_option
+      requested_course_option || current_course_option
+    end
+
+    def current_course_option
+      @application_choice.offered_option
+    end
+
+    def requested_course_option
+      course_option_id = change_offer_params[:course_option_id]
+      CourseOption.find(course_option_id) if course_option_id.present?
+    end
+
+    def set_alternative_providers
+      @alternative_providers = current_provider_user.providers.order(:name)
+    end
+
+    def set_alternative_courses
+      @alternative_courses = Course.where(open_on_apply: true, provider: provider).order(:name)
+    end
+
+    def set_alternative_course_options
+      current_option = @application_choice.offered_option
+      @alternative_course_options = CourseOption.where(
+        course: course,
+        study_mode: current_option.study_mode, # preserving study_mode, for now
+        # TODO: check vacancy_status, e.g. 'B'
+      ).includes(:site).order('sites.name')
+    end
+
+    def change_offer_params
+      begin
+        params.require(:provider_interface_change_offer_form).permit(:provider_id, :course_id, :course_option_id)
```

Comment:
Do we need to rescue? would `params.permit(provider_interface_change_offer_form: [:provider_id, :course_id, :course_option_id])` do the business?

Reply:
No, sorry, doesn't do the trick. I like just doing `change_offer_params[:course_id]` and need to differentiate between a direct navigation to this step (e.g. by clicking to change the course from the application choice page) and a form submission from the previous page (:provider step). This is done partly on the basis of whether the top-level item `provider_interface_changer_offer_form` is present or not.